include/alpaka/mem/BoundaryIter.hpp

```diff
@@ -0,0 +1,395 @@
+        }
+
+        [[nodiscard]] ALPAKA_FN_HOST_ACC constexpr bool isPlane() const
+        {
+            return boundaryDimensionality() == 2;
+        }
+
+        [[nodiscard]] ALPAKA_FN_HOST_ACC constexpr bool isVolume() const
+        {
+            return boundaryDimensionality() == 3;
+        }
+
+        [[nodiscard]] constexpr auto operator<=>(BoundaryDirection const&) const = default;
+    };
+
+    /**
+     * @brief The iterator type for [`BoundaryDirectionsContainer`](@ref)
+     *
+     * @tparam Dim The dimensionality of the volume that this is a boundary direction iterator for.
+     * @tparam LowHaloVecType The vector type used for the lower halo sizes.
+     * @tparam UpHaloVecType The vector type used for the upper halo sizes.
+     */
+    template<uint32_t Dim, alpaka::concepts::Vector LowHaloVecType, alpaka::concepts::Vector UpHaloVecType>
+    struct BoundaryDirectionIter
+    {
+        using BoundaryVecType = alpaka::Vec<BoundaryType, Dim>;
+
+        using difference_type = std::ptrdiff_t;
+        using value_type = BoundaryDirection<Dim, LowHaloVecType, UpHaloVecType>;
+        using reference = value_type&;
+        using const_reference = value_type const&;
+        using pointer = value_type*;
+        using const_pointer = value_type const*;
+
+        ALPAKA_FN_HOST_ACC BoundaryDirectionIter(
+            BoundaryVecType const& boundaries,
+            LowHaloVecType const& lower_halo_sizes,
+            UpHaloVecType const& upper_halo_sizes)
+            : boundaries(boundaries, lower_halo_sizes, upper_halo_sizes)
+            , lowerHaloSizes(lower_halo_sizes)
+            , upperHaloSizes(upper_halo_sizes)
+        {
+        }
+
+        [[nodiscard]] ALPAKA_FN_HOST_ACC constexpr const_reference& operator*() const
+        {
+            return boundaries;
+        }
+
+        [[nodiscard]] ALPAKA_FN_HOST_ACC constexpr reference& operator*()
+        {
+            return boundaries;
+        }
+
+        ALPAKA_FN_HOST_ACC constexpr auto& operator++()
+        {
+            uint32_t i = Dim - 1;
+            bool oob = true;
+            while(i != static_cast<uint32_t>(-1))
+            {
+                switch(boundaries.data[i])
+                {
+                case BoundaryType::LOWER:
+                    boundaries.data[i] = BoundaryType::MIDDLE;
+                    i = static_cast<uint32_t>(-1);
+                    oob = false;
+                    break;
+                case BoundaryType::MIDDLE:
+                    boundaries.data[i] = BoundaryType::UPPER;
+                    i = static_cast<uint32_t>(-1);
+                    oob = false;
+                    break;
+                case BoundaryType::UPPER:
+                    boundaries.data[i] = BoundaryType::LOWER;
+                    --i;
+                    break;
+                case BoundaryType::OOB:
+                    [[fallthrough]];
+                default:
+                    assert(false);
```

Comment:
On the device you should use `ALPAKA_ASSERT_ACC()` the reason is that mostly you would like to disable assertions because they introduce stack frames and influence the kernel performance.

The problem is to figure out if we are within a device or on host. The only idea I have is

```
constexpr bool onHost = bstd::is_same_v<api::Host, ALPAKA_TYPEOF(thisApi())>;
```

// true if we are on the host or within a kernel running on host so we can use `assert)